File: compatibility-validator/app/static/project-map.js
/* Project topology: server-derived ownership, keyboard-accessible SVG and port inventory. */
(() => {
  'use strict';
  const { esc, RequestGate, statusClass } = ValidatorCore;
  const $ = id => document.getElementById(id), gate = new RequestGate();
  const pageSize = 40, portPageSize = 64;
  let context, report = null, page = 0, selected = null, portPage = {}, drawingWidth = 1160;
  const badge = status => `<span class="pill ${statusClass(status)}">${esc(status)}</span>`;
  const short = (value, limit = 32) => String(value ?? '').length > limit ? String(value).slice(0, limit - 1) + '…' : String(value ?? '');
  const modeText = segment => segment.mode ? `${segment.mode.id}: ${segment.mode.links} × ${segment.mode.speed_gbps}G` : `${segment.selected_mode || 'Auto'} — mode unresolved`;
  const checks = rows => `<ul class="map-checks">${rows.map(c => `<li><strong>${esc(c.state)}</strong> ${esc(c.message)}<span class="small"> ${esc(c.code)}</span></li>`).join('')}</ul>`;

  function invalidate() {
    gate.cancel(); report = null; selected = null; page = 0; portPage = {};
    $('mapCanvas').replaceChildren(); $('mapDetails').replaceChildren(); $('mapConnections').replaceChildren(); $('mapChecks').replaceChildren();
    $('mapStatus').textContent = 'Build the map from the current project to inspect connections and physical ports.';
    $('mapPageStatus').textContent = ''; $('buildProjectMap').disabled = false;
    $('mapPrevious').disabled = $('mapNext').disabled = true;
  }
  async function build() {
    let payload;
    try { payload = context.getProject(); } catch (error) { $('mapStatus').textContent = error.message; return; }
    invalidate();
    const ticket = gate.start(context.key());
    const timer = setTimeout(() => ticket.controller.abort(), 30000);
    $('buildProjectMap').disabled = true; $('mapStatus').textContent = 'Validating connections and resolving physical cage capacity…';
    try {
      const response = await fetch('/api/project/map', { method: 'POST', headers: { 'Content-Type': 'application/json' }, cache: 'no-store', signal: ticket.controller.signal, body: JSON.stringify(payload) });
      const value = await response.json();
      if (!gate.current(ticket, context.key())) return;
      if (!response.ok) throw new Error(Array.isArray(value.detail) ? value.detail.slice(0, 5).map(v => `${v.loc.join('.')}: ${v.msg}`).join('; ') : value.detail || 'Map request failed');
      if (value.revision !== payload.revision) throw new Error('Catalog changed. Refresh and build the map again.');
      report = value;
      const current = $('mapDeviceFilter').value;
      $('mapDeviceFilter').innerHTML = '<option value="">All devices</option>' + report.devices.map(d => `<option value="${esc(d.id)}">${esc(d.id)}${d.rack ? ' · ' + esc(d.rack) : ''}</option>`).join('');
      if (report.devices.some(d => d.id === current)) $('mapDeviceFilter').value = current;
      const s = report.summary;
      $('mapStatus').textContent = `${report.status} · ${s.devices} devices · ${s.assemblies} assemblies · ${s.occupied_cages} occupied cages · ${s.conflicting_cages} cage conflicts · catalog ${report.revision}`;
      $('mapChecks').innerHTML = `<details><summary>Project ownership checks</summary>${checks(report.checks)}</details>`;
      render();
    } catch (error) {
      if (ticket.sequence === gate.sequence) $('mapStatus').textContent = error.name === 'AbortError' ? 'Map request timed out. Try again.' : error.message;
    } finally { clearTimeout(timer); if (ticket.sequence === gate.sequence) $('buildProjectMap').disabled = false; }
  }
  function filteredAssemblies() {
    const device = $('mapDeviceFilter').value, fabric = $('mapFabric').value, status = $('mapOutcome').value, query = $('mapSearch').value.trim().toLowerCase();
    return report.assemblies.filter(a => {
      const attached = report.segments.filter(s => s.entry_id === a.id), ids = new Set(attached.map(s => s.instance_id));
      const devices = report.devices.filter(d => ids.has(d.id));
      return (!device || ids.has(device)) && (!fabric || a.fabric === fabric) && (!status || a.status === status) &&
        (!query || [a.id, a.cable_id, a.part_number, ...devices.flatMap(d => [d.id, d.model, d.rack]), ...attached.flatMap(s => [s.part_number, s.port_label])].join(' ').toLowerCase().includes(query));
    });
  }
  function render() {
    if (!report) return;
    const filtered = filteredAssemblies(), pages = Math.max(1, Math.ceil(filtered.length / pageSize));
    page = Math.min(page, pages - 1);
    const assemblies = filtered.slice(page * pageSize, (page + 1) * pageSize), ids = new Set(assemblies.map(a => a.id));
    const segments = report.segments.filter(s => ids.has(s.entry_id)), instances = new Set(segments.map(s => s.instance_id));
    const devices = report.devices.filter(d => instances.has(d.id));
    const leftIds = new Set(segments.filter(s => ['a', 'head'].includes(s.role)).map(s => s.instance_id));
    const left = devices.filter(d => leftIds.has(d.id)), right = devices.filter(d => !leftIds.has(d.id));
    const height = Math.max(340, (Math.max(left.length, right.length, assemblies.length) + 1) * 120);
    const positions = new Map(), assemblyPositions = new Map();
    const distribute = (items, x, width, target) => items.forEach((item, i) => target.set(item.id, { x, y: 30 + i * (height - 100) / Math.max(1, items.length), width }));
    distribute(left, 20, 250, positions); distribute(right, 890, 250, positions); distribute(assemblies, 450, 260, assemblyPositions);
    const isSelected = (kind, id) => selected?.kind === kind && selected.id === id;
    const linked = s => !selected || selected.kind === 'assembly' && selected.id === s.entry_id || selected.kind === 'device' && selected.id === s.instance_id;
    const paths = segments.map(s => {
      const d = positions.get(s.instance_id), a = assemblyPositions.get(s.entry_id), isLeft = d.x < a.x;
      const deviceEnds = segments.filter(v => v.instance_id === s.instance_id), assemblyEnds = segments.filter(v => v.entry_id === s.entry_id);
      const x1 = isLeft ? d.x + d.width : d.x, x2 = isLeft ? a.x : a.x + a.width;
      const y1 = d.y + 12 + 56 * (deviceEnds.indexOf(s) + 1) / (deviceEnds.length + 1), y2 = a.y + 10 + 64 * (assemblyEnds.indexOf(s) + 1) / (assemblyEnds.length + 1);
      const index = report.assemblies.findIndex(v => v.id === s.entry_id);
      const label = `${s.instance_id} / ${s.port_label || s.port_group_id + '/' + s.port_number} → ${s.entry_id} ${s.branch_id || s.role}; ${modeText(s)}; ${s.status}`;
      const middle = (x1 + x2) / 2;
      const branchLabel = s.role === 'branch';
      const labelX = branchLabel ? x1 + (isLeft ? 7 : -7) : x2 + (isLeft ? -7 : 7), labelY = branchLabel ? y1 - 9 : y2 - 7;
      const labelAnchor = branchLabel ? (isLeft ? 'start' : 'end') : (isLeft ? 'end' : 'start');
      return `<g class="map-segment ${linked(s) ? '' : 'map-dim'}" data-assembly="${index}" role="button" tabindex="0" aria-label="${esc(label)}"><title>${esc(label)}</title><path class="map-hit" d="M ${x1} ${y1} C ${middle} ${y1}, ${middle} ${y2}, ${x2} ${y2}"/><path class="map-line state-${s.ownership_conflict ? 'incompatible' : s.status}${s.role === 'head' ? ' map-head' : ''}" d="M ${x1} ${y1} C ${middle} ${y1}, ${middle} ${y2}, ${x2} ${y2}"/><text x="${labelX}" y="${labelY}" text-anchor="${labelAnchor}">${esc(branchLabel && deviceEnds.length > 2 ? '' : s.role === 'branch' ? 'BR ' + s.termination : s.role.toUpperCase())}</text></g>`;
    }).join('');
    const nodes = devices.map(d => {
      const p = positions.get(d.id), index = report.devices.indexOf(d);
      return `<g class="map-device${d.ownership_conflict ? ' map-conflict' : ''}${isSelected('device', d.id) ? ' map-selected' : ''}" data-device="${index}" role="button" tabindex="0" aria-label="${esc('Device ' + d.id + '; ' + d.occupied + ' occupied cages; ' + (d.free_count ?? 'unknown') + ' free')}"><title>${esc(d.id + ' · ' + d.model + (d.rack ? ' · ' + d.rack : ''))}</title><rect x="${p.x}" y="${p.y}" width="${p.width}" height="80" rx="9"/><text x="${p.x + 12}" y="${p.y + 23}" class="map-node-title">${esc(short(d.id))}</text><text x="${p.x + 12}" y="${p.y + 44}">${esc(short(d.model))}</text><text x="${p.x + 12}" y="${p.y + 65}">${esc(d.identity_conflict ? 'Conflicting identity' : `${d.occupied} occupied · ${d.free_count ?? '?'} free cages`)}</text></g>`;
    }).join('');
    const cables = assemblies.map(a => {
      const p = assemblyPositions.get(a.id), index = report.assemblies.indexOf(a);
      return `<g class="map-assembly state-${a.ownership_conflict ? 'incompatible' : a.status}${isSelected('assembly', a.id) ? ' map-selected' : ''}" data-assembly="${index}" role="button" tabindex="0" aria-label="${esc(a.cable_id + '; ' + a.type + '; ' + a.status)}"><title>${esc(a.id + ' · ' + (a.part_number || 'PN unknown'))}</title><rect x="${p.x}" y="${p.y}" width="${p.width}" height="86" rx="18"/><text x="${p.x + 12}" y="${p.y + 23}" class="map-node-title">${esc(short(a.cable_id))}</text><text x="${p.x + 12}" y="${p.y + 43}">${esc(short(a.part_number || 'PN unknown'))}</text><text x="${p.x + 12}" y="${p.y + 64}">${esc(a.ownership_conflict ? 'Port conflict · ' + a.fabric : a.type + ' · ' + a.fabric + ' · ' + a.status)}</text></g>`;
    }).join('');
    $('mapCanvas').innerHTML = assemblies.length ? `<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 ${drawingWidth} ${height}" width="${drawingWidth}" height="${height}" role="group" aria-label="Interactive project connections">${paths}${nodes}${cables}</svg>` : '<p class="empty">No connections match these filters.</p>';
    zoom();
    $('mapPageStatus').textContent = `${filtered.length ? page * pageSize + 1 : 0}–${Math.min((page + 1) * pageSize, filtered.length)} of ${filtered.length} matching assemblies · page ${page + 1}/${pages} · device counts cover the entire project`;
    $('mapPrevious').disabled = page === 0; $('mapNext').disabled = page + 1 >= pages;
    $('mapConnections').innerHTML = assemblies.map(a => `<button type="button" data-assembly="${report.assemblies.indexOf(a)}">${esc(a.cable_id)} · ${esc(a.status)}</button>`).join('');
    renderDetails();
  }
  function zoom() {
    const svg = $('mapCanvas').querySelector('svg');
    if (svg) { const scale = Number($('mapZoom').value) / 100; svg.setAttribute('width', drawingWidth * scale); svg.setAttribute('height', svg.viewBox?.baseVal?.height ? svg.viewBox.baseVal.height * scale : Number(svg.getAttribute('viewBox').split(' ')[3]) * scale); }
    $('mapZoomValue').textContent = $('mapZoom').value + '%';
  }
  function renderDetails() {
    if (!selected) { $('mapDetails').innerHTML = '<p class="small">Select a device to inspect physical cages, or a cable/line to inspect the whole connection and its validation.</p>'; return; }
    if (selected.kind === 'assembly') {
      const a = report.assemblies.find(a => a.id === selected.id);
      if (!a) { selected = null; renderDetails(); return; }
      const segs = report.segments.filter(s => s.entry_id === a.id);
      $('mapDetails').innerHTML = `<h4>${esc(a.cable_id)}</h4><p>${badge(a.status)} ${esc(a.id)} · ${esc(a.type)} · ${esc(a.fabric)}</p>${a.ownership_conflict ? '<p class="bad">Port allocation conflict: review device cages and project ownership checks.</p>' : ''}<p>Cable/harness PN: <strong>${esc(a.part_number || 'unknown')}</strong> · ${a.length_m == null ? 'Length unknown' : esc(a.length_m) + ' m'}</p><p class="small">${a.type === 'breakout' ? 'One shared head cage; each branch is shown separately. Status applies to the complete assembly.' : 'Both physical ends are shown.'}</p><ul>${segs.map(s => `<li><button type="button" data-device="${report.devices.findIndex(d => d.id === s.instance_id)}">${esc(s.instance_id)}</button> ${esc(s.port_group_id)}/${esc(s.port_number)}${s.port_label ? ' · ' + esc(s.port_label) : ''}<div>${esc(s.branch_id || s.role)} · ${esc(modeText(s))}</div><div>Endpoint PN: ${esc(s.part_number || 'unknown')}${s.head_links.length ? ' · head links ' + esc(s.head_links.join(', ')) : ''}</div></li>`).join('')}</ul><details><summary>Branch paths and installation</summary>${a.rows.map(r => `<p>${esc(r.branch_id || 'A ↔ B')} · ${esc(r.source_label)} ↔ ${esc(r.destination_label)}<br>${esc(r.source.instance_id)} → ${esc(r.destination.instance_id)} · installed ${r.installed ? 'yes' : 'no'} · checked ${r.checked ? 'yes' : 'no'}${r.head_optical_lanes.length ? '<br>Optical pairs: ' + esc(r.head_optical_lanes.join(',')) + ' → ' + esc(r.branch_optical_lanes.join(',')) : ''}</p>`).join('')}</details><details open><summary>Validation checks</summary>${checks(a.checks)}</details>`;
      return;
    }
    const d = report.devices.find(d => d.id === selected.id);
    if (!d) { selected = null; renderDetails(); return; }
    $('mapDetails').innerHTML = `<h4>${esc(d.id)}</h4><p>${esc(d.model)}${d.hardware_profile_id ? ' · ' + esc(d.hardware_profile_id) : ''}</p><p>${esc(d.rack || 'Rack unspecified')}${d.rack_u ? ' / U' + esc(d.rack_u) : ''} · ${d.occupied} occupied · ${d.free_count ?? 'unknown'} free physical cages</p>${d.identity_conflict ? '<p class="bad">Conflicting device/profile declarations. Capacity cannot be resolved.</p>' : ''}<p class="small">Free cages are unallocated in this project. Logical interfaces inside an occupied cage do not add free physical ports.</p>${d.groups.map((g, gi) => {
      const pageKey = d.id + '/' + g.id, offset = Math.min(portPage[pageKey] || 0, Math.max(0, Math.ceil((g.count || 1) / portPageSize) - 1));
      const numbers = g.count === null ? g.ports.map(p => p.number) : Array.from({length:Math.min(portPageSize, Math.max(0, g.count - offset * portPageSize))}, (_, i) => i + 1 + offset * portPageSize);
      const extra = g.count === null ? [] : g.ports.filter(p => p.number > g.count).map(p => p.number);
      const ranges = g.free_ranges.map(([a,b]) => a === b ? a : `${a}–${b}`).join(', ');
      return `<section class="map-port-group"><h5>${esc(g.label)} · ${esc(g.connector_family || 'connector unknown')}</h5><p>${g.count ?? 'Unknown'} cages · ${g.occupied} occupied · ${g.free_count ?? 'unknown'} free</p>${g.count === null ? '<p class="warn">Capacity unknown: only declared cages are displayed.</p>' : `<p class="small">Free cage ordinals: ${esc(ranges || 'none')}</p>`}<div class="map-cages">${[...numbers, ...extra].map(n => { const p = g.ports.find(v => v.number === n); return `<button type="button" class="map-cage cage-${p?.state || 'free'}" data-port-group="${gi}" data-port="${n}" aria-label="${esc(g.id + ' cage ' + n + ', ' + (p?.state || 'free'))}">${n}<span>${esc(short(p?.label || p?.state || 'free', 16))}</span></button>`; }).join('')}</div>${g.count > portPageSize ? `<div class="actionbar"><button data-port-page="${gi}" data-offset="${offset - 1}"${offset === 0 ? ' disabled' : ''}>Previous cages</button><span>${offset * portPageSize + 1}–${Math.min(g.count, (offset + 1) * portPageSize)}</span><button data-port-page="${gi}" data-offset="${offset + 1}"${(offset + 1) * portPageSize >= g.count ? ' disabled' : ''}>Next cages</button></div>` : ''}</section>`;
    }).join('')}<div id="mapPortDetail" aria-live="polite"><p>Select a cage to inspect its allocation and logical interfaces.</p></div>`;
  }
  function selectPort(groupIndex, number) {
    const d = report.devices.find(d => d.id === selected.id), g = d.groups[groupIndex], port = g.ports.find(p => p.number === number);
    const segs = port ? report.segments.filter(s => port.segments.includes(s.id)) : [];
    $('mapPortDetail').innerHTML = `<h5>${esc(g.id)} / cage ${number}${port?.label ? ' · ' + esc(port.label) : ''}</h5><p>${esc(port?.state || 'free in this project')}</p>${port ? `<ul>${segs.map(s => `<li><button type="button" data-assembly="${report.assemblies.findIndex(a => a.id === s.entry_id)}">${esc(s.entry_id)} · ${esc(s.branch_id || s.role)}</button><div>${esc(modeText(s))}</div></li>`).join('')}</ul><h5>Logical interfaces within this physical cage</h5>${port.interfaces.length ? `<ul>${port.interfaces.map(i => `<li>Interface ${i.number} · ${i.speed_gbps}G · ${esc(i.state)}${i.references.length ? ' → ' + esc(i.references.join(', ')) : ''}</li>`).join('')}</ul>` : '<p class="warn">Logical interfaces unresolved: choose an explicit supported mode and resolve cage conflicts.</p>'}` : '<p>No cable is allocated here in this project; the physical port is not a live discovery result.</p>'}`;
  }
  function activate(event) {
    if (!report) return;
    const target = event.target.closest('[data-device],[data-assembly],[data-port],[data-port-page]');
    if (!target) return;
    if (event.type === 'keydown') { if (!['Enter', ' '].includes(event.key) || target.tagName.toLowerCase() === 'button') return; event.preventDefault(); }
    if (target.dataset.port !== undefined) { selectPort(Number(target.dataset.portGroup), Number(target.dataset.port)); return; }
    if (target.dataset.portPage !== undefined) {
      const d = report.devices.find(d => d.id === selected.id), g = d.groups[Number(target.dataset.portPage)];
      portPage[d.id + '/' + g.id] = Number(target.dataset.offset); renderDetails(); return;
    }
    const kind = target.dataset.device !== undefined ? 'device' : 'assembly';
    const item = kind === 'device' ? report.devices[Number(target.dataset.device)] : report.assemblies[Number(target.dataset.assembly)];
    selected = { kind, id: item.id }; render();
    if (event.type === 'keydown') $('mapDetails').focus();
  }
  $('buildProjectMap').addEventListener('click', build);
  for (const id of ['mapDeviceFilter', 'mapFabric', 'mapOutcome', 'mapSearch']) $(id).addEventListener(id === 'mapSearch' ? 'input' : 'change', () => { page = 0; selected = null; render(); });
  $('mapPrevious').addEventListener('click', () => { page--; render(); });
  $('mapNext').addEventListener('click', () => { page++; render(); });
  $('mapZoom').addEventListener('input', zoom);
  $('mapReset').addEventListener('click', () => { for (const id of ['mapDeviceFilter', 'mapFabric', 'mapOutcome', 'mapSearch']) $(id).value = ''; $('mapZoom').value = '100'; page = 0; selected = null; render(); $('mapCanvas').scrollTop = $('mapCanvas').scrollLeft = 0; });
  for (const id of ['mapCanvas', 'mapDetails', 'mapConnections']) { $(id).addEventListener('click', activate); $(id).addEventListener('keydown', activate); }
  window.addEventListener('pagehide', () => gate.cancel());
  window.ValidatorProjectMap = { init: value => { context = value; }, invalidate, renderDraft: project => { $('buildProjectMap').disabled = !project.connections.length && !project.breakouts.length; } };
})();

File: compatibility-validator/app/project_map.py
"""Derived topology and physical cage inventory for one validated project."""

from collections import defaultdict

from .cabling import build_cabling
from .hardware import effective_host
from .models import HostSelection


def free_ranges(count, occupied):
    """Describe unused ordinals without allocating a catalog-sized array."""
    if count is None:
        return []
    result, start = [], 1
    for number in sorted(n for n in occupied if 1 <= n <= count):
        if start < number:
            result.append([start, number - 1])
        start = number + 1
    if start <= count:
        result.append([start, count])
    return result


def build_project_map(snapshot, project, validation):
    plan = build_cabling(snapshot, project, validation)
    results = {r['id']: r for r in validation['results']}
    locations = {p.instance_id: p for p in project.cabling.locations}
    markings = {(p.instance_id, p.port_group_id, p.port_number): p.label for p in project.cabling.port_labels}
    models = {d['id']: d for d in snapshot['devices']}
    profiles = {p['id']: p for p in snapshot.get('hardware_profiles', [])}
    selections, segments, assemblies = defaultdict(list), [], []

    def segment(entry, host, role, branch=None):
        try:
            _, group, _ = effective_host(snapshot, host)
            mode = next((m for m in group['modes'] if m['id'] == host.mode_id), None)
        except KeyError:
            mode = None
        item = dict(id=f'segment-{len(segments)}', entry_id=entry.id, instance_id=host.instance_id,
            port_group_id=host.port_group_id, port_number=host.port_number,
            port_label=markings.get((host.instance_id, host.port_group_id, host.port_number)),
            role=role, branch_id=branch.id if branch else None, termination=branch.termination if branch else None,
            head_links=branch.head_links if branch else [], selected_mode=host.mode_id,
            mode={k: mode[k] for k in ('id', 'links', 'speed_gbps')} if mode else None,
            part_number=host.part_number, status=results[entry.id]['status'])
        segments.append(item)
        selections[host.instance_id].append((host, item, entry))

    for entry in [*project.connections, *project.breakouts]:
        is_breakout = hasattr(entry, 'branches')
        rows = [r for r in plan['rows'] if r['entry_id'] == entry.id]
        result = results[entry.id]
        assemblies.append(dict(id=entry.id, type='breakout' if is_breakout else 'connection',
            topology=entry.topology if is_breakout else 'point-to-point', cable_id=rows[0]['cable_id'],
            part_number=rows[0]['part_number'], length_m=rows[0]['length_m'], fabric=entry.fabric,
            status=result['status'], checks=result['validation']['checks'], rows=rows))
        if is_breakout:
            segment(entry, entry.head, 'head')
            for branch in entry.branches:
                segment(entry, branch.selection, 'branch', branch)
        else:
            segment(entry, entry.a, 'a')
            segment(entry, entry.b, 'b')

    devices = []
    for instance, assignments in sorted(selections.items()):
        signatures = {(h.device_id, h.hardware_profile_id) for h, _, _ in assignments}
        identity_conflict = len(signatures) != 1
        first = assignments[0][0]
        device = models.get(first.device_id)
        profile = profiles.get(first.hardware_profile_id) if first.hardware_profile_id else None
        profile_valid = not first.hardware_profile_id or bool(profile and profile['device_id'] == first.device_id)
        base_groups = {g['id']: g for g in device['port_groups']} if device else {}
        allowed = {p['port_group_id'] for p in profile['ports']} if profile and profile_valid else set(base_groups)
        group_ids = allowed | {h.port_group_id for h, _, _ in assignments}
        groups = []
        for group_id in sorted(group_ids):
            base = base_groups.get(group_id, {})
            group = None
            if not identity_conflict and profile_valid:
                try:
                    _, group, _ = effective_host(snapshot, HostSelection(device_id=first.device_id,
                        hardware_profile_id=first.hardware_profile_id, port_group_id=group_id))
                except KeyError:
                    pass
            count = group.get('count') if group else None
            port_uses = defaultdict(list)
            for host, item, entry in assignments:
                if host.port_group_id == group_id:
                    port_uses[host.port_number].append((host, item, entry))
            ports = []
            for number, uses in sorted(port_uses.items()):
                ownership_conflict = identity_conflict or len(uses) > 1 or count is not None and number > count
                for _, item, _ in uses:
                    item['ownership_conflict'] = ownership_conflict
                interfaces = []
                if len(uses) == 1 and uses[0][1]['mode']:
                    _, item, entry = uses[0]
                    for logical in range(1, item['mode']['links'] + 1):
                        if item['role'] == 'head':
                            references = [b.id for b in entry.branches if logical in b.head_links]
                        elif item['mode']['links'] == 1:
                            references = [item['branch_id'] or entry.id]
                        else:
                            references = []
                        interfaces.append(dict(number=logical, speed_gbps=item['mode']['speed_gbps'],
                            references=references, state='conflict' if len(references) > 1 else 'declared' if references else 'unmapped'))
                ports.append(dict(number=number, label=markings.get((instance, group_id, number)),
                    state='conflict' if len(uses) > 1 else 'out-of-range' if count is not None and number > count else 'occupied',
                    segments=[item['id'] for _, item, _ in uses], interfaces=interfaces))
            in_range = sum(1 for n in port_uses if count is not None and n <= count)
            groups.append(dict(id=group_id, label=base.get('label', group_id),
                connector_family=base.get('connector_family'), count=count,
                occupied=len(ports), occupied_in_range=in_range if count is not None else None,
                free_count=count - in_range if count is not None else None,
                free_ranges=free_ranges(count, port_uses), ports=ports,
                capacity_known=count is not None))
        location = locations.get(instance)
        devices.append(dict(id=instance, device_id=first.device_id,
            model=device['model'] if device and not identity_conflict else 'Conflicting models/profiles' if identity_conflict else first.device_id,
            hardware_profile_id=first.hardware_profile_id, identity_conflict=identity_conflict,
            ownership_conflict=any(s.get('ownership_conflict') for _, s, _ in assignments),
            rack=location.rack if location else None, rack_u=location.rack_u if location else None,
            groups=groups, occupied=sum(g['occupied'] for g in groups),
            free_count=sum(g['free_count'] for g in groups) if groups and all(g['capacity_known'] for g in groups) else None))
    for assembly in assemblies:
        assembly['ownership_conflict'] = any(s.get('ownership_conflict') for s in segments if s['entry_id'] == assembly['id'])
    return dict(scope='project-map', name=project.name, revision=snapshot['revision'], status=validation['status'],
        devices=devices, assemblies=assemblies, segments=segments,
        checks=[c for c in validation['checks'] if c['code'].startswith(('ports.', 'project.'))],
        summary=dict(devices=len(devices), assemblies=len(assemblies), terminations=len(segments),
            occupied_cages=sum(d['occupied'] for d in devices),
            conflicting_cages=sum(p['state'] == 'conflict' for d in devices for g in d['groups'] for p in g['ports'])),
        notes=['Free means unallocated in this project; live device occupancy is not discovered.',
            'Physical cage counts come from the selected exact profile or catalog. Unknown or conflicting identities never imply free capacity.',
            'Logical interfaces belong to a selected cage mode and are not additional physical cages. Unmapped interfaces are not spare cages.',
            'Cable and breakout status comes from complete validation. A physical cage remains occupied even when compatibility fails.'])
